### PR TITLE
Support unique client id in MoP cluster.

### PR DESCRIPTION
**Pull request by Technoboy-**
Master Issue: #409

### Motivation

As #560 has introduced system topic, this patch is supporting unique client id in MoP cluster.

### Documentation

- [x] `doc-required` 

Update related doc together.

### CODE REVIEW

**Comment by mattisonchao** [comment]
@Technoboy-  Great work!

**Review by mattisonchao — COMMENTED** [review]
Maybe we need to be compatible with MQTT version 5 to send a ``disconnect`` message to the client to avoid endless reconnection.

File: mqtt-impl/src/main/java/io/streamnative/pulsar/handlers/mqtt/support/systemtopic/SystemTopicBasedSystemEventService.java [hunk 1]
```diff
@@ -63,27 +61,21 @@ public SystemTopicBasedSystemEventService(PulsarService pulsarService) {
...
-        return ip != null && !ip.equals(NettyUtils.getIp(connection.getChannel()));
-    }
-
     @Override
     public CompletableFuture<Void> sendConnectEvent(Connection connection) {
+        checkReader();
         ConnectEvent connectEvent = ConnectEvent.builder()
                 .clientId(connection.getClientId())
-                .ip(NettyUtils.getIp(connection.getChannel()))
+                .address(pulsarService.getAdvertisedAddress())
```
Comment (mattisonchao) [inline mqtt-impl/src/main/java/io/streamnative/pulsar/handlers/mqtt/support/systemtopic/SystemTopicBasedSystemEventService.java:78]: I'm not sure about this, why we don't use client ip+port?

**Review by mattisonchao — APPROVED** [review]

File: mqtt-impl/src/main/java/io/streamnative/pulsar/handlers/mqtt/proxy/MQTTProxyProtocolMethodProcessor.java [hunk 2]
```diff
@@ -98,7 +101,11 @@ public void doProcessConnect(MqttConnectMessage msg, String userRole, ClientRest
                 .channel(channel)
                 .connectionManager(connectionManager)
                 .build();
-        connection.sendConnAck();
+        connection.sendConnAck().addListener(listener -> {
+            if (listener.isSuccess()) {
+                eventService.sendConnectEvent(connection);
+            }
```
[the remark this answers is not in the document]
Comment (Technoboy-) [reply mqtt-impl/src/main/java/io/streamnative/pulsar/handlers/mqtt/proxy/MQTTProxyProtocolMethodProcessor.java:107]: connection.sendConnAck() will log when error. So I think it's no need to log here.

File: mqtt-impl/src/main/java/io/streamnative/pulsar/handlers/mqtt/MQTTConnectionManager.java [hunk 3]
```diff
@@ -82,4 +93,29 @@ public void removeConnection(Connection connection) {
...
+
+        @Override
+        public void onChange(MqttEvent event) {
+            if (event.getEventType() == CONNECT) {
+                ConnectEvent connectEvent = (ConnectEvent) event.getSourceEvent();
+                if (!connectEvent.getAddress().equals(advertisedAddress)) {
+                    Connection connection = getConnection(connectEvent.getClientId());
+                    if (connection != null) {
+                        if (log.isDebugEnabled()) {
+                            log.debug("[ConnectEvent] close existing connection : {}", connection);
```
[the remark this answers is not in the document]
Comment (Technoboy-) [reply mqtt-impl/src/main/java/io/streamnative/pulsar/handlers/mqtt/MQTTConnectionManager.java:108]: Right.

[the remark this answers is not in the document]
Comment (Technoboy-) [reply mqtt-impl/src/main/java/io/streamnative/pulsar/handlers/mqtt/MQTTConnectionManager.java]: fixed

**Comment by Technoboy-** [comment]
> Maybe we need to be compatible with MQTT version 5 to send a `disconnect` message to the client to avoid endless reconnection.

It's better to obey the suggestion that given by MQTT spec. Only override connection.

**Comment by mattisonchao** [comment]
@Technoboy-  
In MQTT 5.0 spec they define we need to send disconnect to client.

> If the ClientID represents a Client already connected to the Server, the Server sends a DISCONNECT packet to the existing Client with Reason Code of 0x8E (Session taken over) as described in [section 4.13](https://docs.oasis-open.org/mqtt/mqtt/v5.0/os/mqtt-v5.0-os.html#S4_13_Errors) and MUST close the Network Connection of the existing Client [MQTT-3.1.4-3]. If the existing Client has a Will Message, that Will Message is published as described in section 3.1.2.5.

**Comment by Technoboy-** [comment]
> @Technoboy- In MQTT 5.0 spec they define we need to send `disconnect` message to client.
> 
> > If the ClientID represents a Client already connected to the Server, the Server sends a DISCONNECT packet to the existing Client with Reason Code of 0x8E (Session taken over) as described in [section 4.13](https://docs.oasis-open.org/mqtt/mqtt/v5.0/os/mqtt-v5.0-os.html#S4_13_Errors) and MUST close the Network Connection of the existing Client [MQTT-3.1.4-3]. If the existing Client has a Will Message, that Will Message is published as described in section 3.1.2.5.

Ah, ok.

**Comment by Technoboy-** [comment]
> If the ClientID represents a Client already connected to the Server

Fixed.

**Pull request merged**